File: longestPalidrome.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Solution {
    public int longestPalindrome(String s) {

        if(s.length() == 1)
        {
            return 1;
        }

        int count[] = new int[128];
        int ans = 0;
        boolean oddIncluded = false;

        for(char ch: s.toCharArray())
        {
            count[ch]++;
        }

        for(int i = 0 ; i<count.length ; i++)
        {
            if(count[i] == 0)
            {
                continue;
            }
            if(count[i]%2 == 0)
            {
                ans+= count[i];
            }
            else
            {
                if(!oddIncluded)
                {
                    ans+=count[i];
                    oddIncluded = true;
                }
                else
                {
                    ans+=count[i]-1;
                }
            }
        }
        return ans;
    }
}

File: LongestPalidrome.py
"""
Time Complexity : O(n)
Space Complexity : O(1)
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this: none
"""

class Solution:
    def longestPalindrome(self, s):
        
        if len(s) == 0:
            return 0
        
        letter_map = {}
        tot_len = 0
        odd = 0
        
        for i in s:
            if i not in letter_map:
                letter_map[i] = 1
            else:
                letter_map[i] += 1
        
        for k,v in letter_map.items():
            if v % 2 == 0:
                tot_len += v
            else:
                tot_len += v - 1
                odd = 1
        
        return tot_len + odd
        
s = Solution()
print(s.longestPalindrome("abccccdd"))
print(s.longestPalindrome("civilwartestingwhetherthatnaptionoranynartionsoconceivedandsodedicatedcanlongendureWeareqmetonagreatbattlefiemldoftzhatwarWehavecometodedicpateaportionofthatfieldasafinalrestingplaceforthosewhoheregavetheirlivesthatthatnationmightliveItisaltogetherfangandproperthatweshoulddothisButinalargersensewecannotdedicatewecannotconsecratewecannothallowthisgroundThebravelmenlivinganddeadwhostruggledherehaveconsecrateditfaraboveourpoorponwertoaddordetractTgheworldadswfilllittlenotlenorlongrememberwhatwesayherebutitcanneverforgetwhattheydidhereItisforusthelivingrathertobededicatedheretotheulnfinishedworkwhichtheywhofoughtherehavethusfarsonoblyadvancedItisratherforustobeherededicatedtothegreattdafskremainingbeforeusthatfromthesehonoreddeadwetakeincreaseddevotiontothatcauseforwhichtheygavethelastpfullmeasureofdevotionthatweherehighlyresolvethatthesedeadshallnothavediedinvainthatthisnationunsderGodshallhaveanewbirthoffreedomandthatgovernmentofthepeoplebythepeopleforthepeopleshallnotperishfromtheearth"))